File: api-ai-medicare-assistant/AI.MedicareAssistant.Application/Services/PlanCardEnrichmentService.cs
using Application.DTOs;
using Domain.Models;

namespace Application.Services;

/// <summary>
/// Computes derived display fields for plan recommendation cards.
/// Pure computation — no I/O, no DI.
/// </summary>
public static class PlanCardEnrichmentService
{
    public static EnrichedPartDCard EnrichPartD(
        RecommendationListItem plan,
        PartDPlanRecommendationResponse response,
        List<string> selectedPharmacyNumbers,
        int totalDrugs)
    {
        var rec = plan.PharmacyWiseRecommendations.FirstOrDefault();
        var networkCount = rec?.PharmacyNetworks
            .Count(pn => selectedPharmacyNumbers.Contains(pn.PharmacyNumber)) ?? 0;
        var unavailableCount = rec?.UnavailableDrugs?.Count ?? 0;

        return new EnrichedPartDCard
        {
            PlanIdDisplay = $"{plan.ContractId}-{plan.PlanId}-{plan.SegmentId}",
            InsuranceCarrier = response.ContractIdCarrierMap.TryGetValue(plan.ContractId, out var carrier)
                ? carrier : "",
            PartDSurcharge = response.PartDPremiumSurcharge,
            PrescriptionOOP = rec?.TotalPrescriptionCost ?? 0,
            PharmaciesInNetwork = networkCount,
            TotalSelectedPharmacies = selectedPharmacyNumbers.Count,
            DrugsCovered = totalDrugs - unavailableCount,
            TotalDrugs = totalDrugs,
        };
    }

    public static EnrichedMedigapCard EnrichMedigap(
        MedigapPlanQuote quote,
        MedigapPlanQuotesResponse response)
    {
        var carrierName = "";
        if (quote.Naic != null && response.ContractIdCarrierMap != null)
        {
            response.ContractIdCarrierMap.TryGetValue(quote.Naic, out carrierName);
            carrierName ??= "";
        }

        return new EnrichedMedigapCard
        {
            PremiumMonthly = (double)(quote.Rate?.Month ?? 0) / 100.0,
            PremiumAnnual = (double)(quote.Rate?.Annual ?? 0) / 100.0,
            InsuranceCarrier = carrierName,
            PartBSurcharge = (double)(quote.PartBPremiumSurcharge ?? 0),
            HealthcareOOP = (double)(quote.PartBServiceOOP ?? 0),
            RemainingMonths = quote.MonthsUsedForExpenseCalc ?? 0,
        };
    }

    public static EnrichedMACard EnrichMA(
        RecommendationListItem plan,
        PartDPlanRecommendationResponse response,
        List<string> selectedPharmacyNumbers,
        int totalDrugs)
    {
        var rec = plan.PharmacyWiseRecommendations.FirstOrDefault();
        var surcharges = response.PartBPremiumSurcharge + response.PartDPremiumSurcharge;
        var networkCount = rec?.PharmacyNetworks
            .Count(pn => selectedPharmacyNumbers.Contains(pn.PharmacyNumber)) ?? 0;
        var unavailableCount = rec?.UnavailableDrugs?.Count ?? 0;

        return new EnrichedMACard
        {
            PlanIdDisplay = $"{plan.ContractId}-{plan.PlanId}-{plan.SegmentId}",
            InsuranceCarrier = response.ContractIdCarrierMap.TryGetValue(plan.ContractId, out var carrier)
                ? carrier : "",
            Surcharges = surcharges,
            PrescriptionOOP = rec?.TotalPrescriptionCost ?? 0,
            HealthcareOOP = rec?.PartAandBBenefitServiceCost ?? 0,
            HasPrescriptionDrug = rec?.PrescriptionDrugCovered ?? false,
            PharmaciesInNetwork = networkCount,
            TotalSelectedPharmacies = selectedPharmacyNumbers.Count,
            DrugsCovered = totalDrugs - unavailableCount,
            TotalDrugs = totalDrugs,
        };
    }
}

File: ui-ai-medicare-assistant/src/app/services/plan-card-enrichment.service.ts
import { Injectable } from '@angular/core';
import {
  PartDPlanRecommendationResponse,
  RecommendationListItem,
  EnrichedPartDCard,
  EnrichedMACard,
} from '../models/part-d-plan.model';
import {
  MedigapPlanQuotesResponse,
  MedigapPlan,
  EnrichedMedigapCard,
} from '../models/medigap-plan.model';

/**
 * Pure-computation service that derives display fields for plan cards
 * from raw API responses. No HTTP calls.
 */
@Injectable({ providedIn: 'root' })
export class PlanCardEnrichmentService {

  enrichPartD(
    plan: RecommendationListItem,
    response: PartDPlanRecommendationResponse,
    selectedPharmacyNumbers: string[],
    totalDrugs: number,
  ): EnrichedPartDCard {
    const rec = plan.pharmacyWiseRecommendations?.[0];
    const networkCount = rec?.pharmacyNetworks
      ?.filter(pn => selectedPharmacyNumbers.includes(pn.pharmacyNumber)).length ?? 0;
    const unavailable = rec?.unavailableDrugs?.length ?? 0;

    return {
      planIdDisplay: `${plan.contractId}-${plan.planId}-${plan.segmentId}`,
      insuranceCarrier: response.contractIdCarrierMap?.[plan.contractId] ?? '',
      partDSurcharge: response.partDPremiumSurcharge ?? 0,
      prescriptionOOP: rec?.totalPrescriptionCost ?? 0,
      pharmaciesInNetwork: networkCount,
      totalSelectedPharmacies: selectedPharmacyNumbers.length,
      drugsCovered: totalDrugs - unavailable,
      totalDrugs,
    };
  }

  enrichMedigap(
    quote: MedigapPlan,
    response: MedigapPlanQuotesResponse,
  ): EnrichedMedigapCard {
    const carrierName = (quote.naic && response.contractIdCarrierMap)
      ? (response.contractIdCarrierMap[quote.naic] ?? '')
      : '';

    return {
      premiumMonthly: (quote.rate?.month ?? 0) / 100,
      premiumAnnual: (quote.rate?.annual ?? 0) / 100,
      insuranceCarrier: carrierName,
      partBSurcharge: quote.partBPremiumSurcharge ?? 0,
      healthcareOOP: quote.partBServiceOOP ?? 0,
      remainingMonths: quote.monthsUsedForExpenseCalc ?? 0,
    };
  }

  enrichMA(
    plan: RecommendationListItem,
    response: PartDPlanRecommendationResponse,
    selectedPharmacyNumbers: string[],
    totalDrugs: number,
  ): EnrichedMACard {
    const rec = plan.pharmacyWiseRecommendations?.[0];
    const surcharges = (response.partBPremiumSurcharge ?? 0) + (response.partDPremiumSurcharge ?? 0);
    const networkCount = rec?.pharmacyNetworks
      ?.filter(pn => selectedPharmacyNumbers.includes(pn.pharmacyNumber)).length ?? 0;
    const unavailable = rec?.unavailableDrugs?.length ?? 0;

    return {
      planIdDisplay: `${plan.contractId}-${plan.planId}-${plan.segmentId}`,
      insuranceCarrier: response.contractIdCarrierMap?.[plan.contractId] ?? '',
      surcharges,
      prescriptionOOP: rec?.totalPrescriptionCost ?? 0,
      healthcareOOP: rec?.partAandBBenefitServiceCost ?? 0,
      hasPrescriptionDrug: rec?.prescriptionDrugCovered ?? false,
      pharmaciesInNetwork: networkCount,
      totalSelectedPharmacies: selectedPharmacyNumbers.length,
      drugsCovered: totalDrugs - unavailable,
      totalDrugs,
    };
  }
}
